qiskit/transpiler/passes/routing/bip_mapping.py

```diff
@@ -0,0 +1,229 @@
+
+                * ``'error_rate'``: [NotImplemented] Predicted error rate of the circuit
+                * ``'depth'``: [Default] Depth (number of time-steps) of the circuit
+                * ``'balanced'``: [NotImplemented] Weighted sum of ``'error_rate'`` and ``'depth'``
+
+            backend_prop (BackendProperties): Backend properties object
+            time_limit (float): Time limit for solving BIP in seconds
+            threads (int): Number of threads to be allowed for CPLEX to solve BIP
+            max_swaps_inbetween_layers (int):
+                Number of swaps allowed inbetween layers. If None, automatically set.
+                Large value could decrease the probability to build infeasible BIP problem but also
+                could reduce the chance of finding a feasible solution within the ``time_limit``.
+
+        Raises:
+            MissingOptionalLibraryError: if cplex is not installed.
+        """
+        if not HAS_DOCPLEX or not HAS_CPLEX:
+            raise MissingOptionalLibraryError(
+                libname="bip-mapper",
+                name="BIP-based mapping pass",
+                pip_install="pip install 'qiskit-terra[bip-mapper]'",
+            )
+        super().__init__()
+        self.coupling_map = copy.deepcopy(coupling_map)  # save a copy to modify
+        if self.coupling_map is not None:
+            self.coupling_map.make_symmetric()
+        self.objective = objective
+        self.backend_prop = backend_prop
+        self.time_limit = time_limit
+        self.threads = threads
+        self.max_swaps_inbetween_layers = max_swaps_inbetween_layers
+
+    def run(self, dag):
+        """Run the BIPMapping pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): DAG to map.
+
+        Returns:
+            DAGCircuit: A mapped DAG. If failing to map, returns the original dag.
+
+        Raises:
+            TranspilerError: if it fails to solve BIP problem within the time limit
+        """
+        if self.coupling_map is None:
+            return dag
+
+        if len(dag.qubits) > self.coupling_map.size():
+            raise TranspilerError("More virtual qubits exist than physical qubits.")
+
+        if self.property_set["layout"]:
+            logger.info("BIPMapping ignores given initial layout.")
+
+        dummy_steps = math.ceil(math.sqrt(dag.num_qubits()))
+        if self.max_swaps_inbetween_layers is not None:
+            dummy_steps = max(0, self.max_swaps_inbetween_layers - 1)
+
+        original_dag = dag
+        # BIPMappingModel assumes num_virtual_qubits == num_physical_qubits
+        # TODO: rewrite without dag<->circuit conversion (or remove the above assumption)
+        pm = PassManager(
+            [
+                TrivialLayout(self.coupling_map),
+                FullAncillaAllocation(self.coupling_map),
+                EnlargeWithAncilla(),
+            ]
+        )
+        dag = circuit_to_dag(pm.run(dag_to_circuit(dag)))
```

Comment:
I dont like running a passmanager from within a pass. Can't this just be handled as a prerequisite preceding pass? Isn't this how Sabre, etc work?

Reply:
A lot of passes have non-obvious preconditions that must be satisfied by the external flow. I would like to do more about that by checking/asserting pre- and post-conditions for passes, but I don't think we need to solve it in this PR. Raising errors for easily-checkable preconditions is a good start though.

I agree that the semantics of `initial_layout` can be confusing and we should (in a separate PR) separate out the meanings which currently seem to include "I require the input qubits to be mapped this way", "here is a suggestion for how to map the input qubits, but if you can find a better one use it", "use this mapping for input qubits if and only if it complies with the coupling map" and probably some others. (We also need to be clear about what happens with the mapping of output qubits)

Reply:
Passes can declare that they require some other passes to be run before them. They can do this via the `requires` attribute. See example here: https://github.com/Qiskit/qiskit-terra/blob/06550512e30436c7bc9bbcf54e647694e198ac9a/qiskit/transpiler/passes/scheduling/alap.py#L36

If the pass is invoked individually, this will be ignored. But if it is invoked by the PassManager, then the PassManager will ensure that these requirements are met (which is why invoking passes individually is usually discouraged).

Is that enough to solve the problem here?


> we cannot check if a user layout is ignored or not within the BIPMapping pass and it has to be cared in the external flow

Yes I think that's the correct way to do this. I think this kind of thing already happens in the PassManager, none of the other passes know whether the layout was given by the user or computed by some other layout pass.

Reply:
Thank you for the comments. I've updated the code accordingly at c2660af and 60d1be9.

Reply:
> I agree that the semantics of `initial_layout` can be confusing and we should (in a separate PR) separate out the meanings which currently seem to include "I require the input qubits to be mapped this way", "here is a suggestion for how to map the input qubits, but if you can find a better one use it", "use this mapping for input qubits if and only if it complies with the coupling map" and probably some others. (We also need to be clear about what happens with the mapping of output qubits)

Just for posterity, the second interpretation here (`initial_layout` is a suggestion, but if it you can find better, use that instead) is wrong, or at least entirely inconsistent with the way in which we handle arguments passed to the transpiler elsewhere. If a user-supplied argument to `transpile` is ignored or overridden elsewhere, that's a bug.

Reply:
Thank you for correcting me, @kdk . I reread preset pass managers and recognized I had misread the spec of `transpile` arguments. I had thought `initial_layout` is sometimes overridden but it's never as you pointed out (`layout_method` is ignored if `initial_layout` is supplied instead).